fission-core/library/Fission/Text/Encoded/Types.hs

```diff
@@ -0,0 +1,35 @@
+module Fission.Text.Encoded.Types (Encoded (..)) where
+
+import           Data.Kind
+
+import           RIO
+import qualified RIO.ByteString.Lazy as Lazy
+import qualified RIO.Text as Text
+
+import           Fission.Text.Encoding.Types
+
+-- | Tag some text representation with the base it's encoded with
+--   NOTE that this is by no means a perfect method, and is mostly to make
+--   tracing encoding through code less painful. Character encoding can always
+--   be forgotten, mislabelled, and so on.
+newtype (enc :: Encoding) `Encoded` (carrierText :: Type)
+  = Encoded { encoded :: carrierText }
+  deriving stock   (Eq, Ord, Show)
+  deriving newtype (Semigroup, Monoid)
```

Comment:
Wait, `carrierText :: Type`? There's some light type-level stuff happening here. This is all to keep the checks at compile-time, prevent accidentally concatenating text in different bases, and so on.

If there's interest in learning about type-level programming, we can have a lunch & learn. In the meantime, just can absolutely use it as a magic interface that looks like:

```haskell
foo :: 'Base58_BTC `Encoded` Text
foo = Encoded "abcdef"

bar :: 'UTF8 `Encoded` Text
bar = Encoded "12345"

foo <> foo -- Encoded "abcdefabcdef"
foo <> bar -- Compiler error, Because 'UTF8 is not 'Base58_BTC
```